src/extra/libs/rlottie/rlottiedec.c

```diff
@@ -0,0 +1,333 @@
+        dec_ctx->cache = animation;
+    }
+    else {
+        header->w = (uint32_t)w;
+        header->h = (uint32_t)h;
+        lottie_animation_destroy(animation);
+    }
+    header->always_zero = 0;
+    header->cf = LV_IMG_CF_TRUE_COLOR_ALPHA;
+    return LV_RES_OK;
+}
+
+
+/**
+ * Open a rlottie animation image and return the decoded image
+ * @param dsc Decoded descriptor for the animation
+ * @return LV_RES_OK: no error; LV_RES_INV: can't decode the picture
+ */
+static lv_res_t decoder_open(lv_img_decoder_t * decoder, lv_img_decoder_dsc_t * dsc, void * _dec_ctx)
+{
+    LV_UNUSED(decoder);
+    rlottiedec_ctx_t * dec_ctx = (rlottiedec_ctx_t *)_dec_ctx;
+
+    Lottie_Animation * animation = NULL;
+    size_t w, h;
+    /* If already opened, reuse it */
+    if(dec_ctx != NULL && dec_ctx->ctx.magic_id == RLOTTIE_ID && dec_ctx->cache != NULL)
+        animation = dec_ctx->cache;
+    /*If it's a rlottie JSON file...*/
+    else if(dsc->src_type == LV_IMG_SRC_FILE) {
+        const char * fn = dsc->src;
+
+        if(strcmp(&fn[strlen(fn) - 4], "json")) return LV_RES_INV;       /*Check the extension*/
+        animation = lottie_animation_from_file(fn);
+    }
+    /* rlottie as raw data */
+    else if(dsc->src_type == LV_IMG_SRC_VARIABLE) {
+        const char * fn = dsc->src;
+        animation = lottie_animation_from_data(fn, fn, "");
+    }
+    if(animation == NULL)
+        return LV_RES_INV;
+
+    lv_rlottie_dec_context_t * context = lv_mem_alloc(sizeof(lv_rlottie_dec_context_t));
+    if(dec_ctx == NULL) {
+        lottie_animation_get_size(animation, &w, &h);
+        dec_ctx = lv_mem_alloc(sizeof(rlottiedec_ctx_t));
+        dec_ctx->create_width = w;
+        dec_ctx->create_height = h;
+        dec_ctx->should_free = 1;
+    }
+
+    dec_ctx->ctx.total_frames = lottie_animation_get_totalframe(animation);
+    dec_ctx->cache = animation;
+    w = dec_ctx->create_width;
+    h = dec_ctx->create_height;
+    context->dec_ctx = dec_ctx;
+
+    /* Compute how many lines we can fit in the maximum buffer size */
+    context->scanline_width = w * LV_ARGB32 / 8;
+    context->lines_in_buf = max_buf_size / context->scanline_width;
+    if(context->lines_in_buf != 0) {  /*Too big picture to fit the maximum buffer size (default to 1024px) ?*/
+        context->allocated_buf = lv_mem_alloc(context->lines_in_buf * context->scanline_width);
+    }
+    if(context->allocated_buf == NULL) {
+        if(dec_ctx->should_free) {
+            lottie_animation_destroy(animation);
+            dec_ctx->cache = 0;
+            lv_mem_free(dec_ctx);
+        }
+        return LV_RES_INV;
+    }
+
+    memset(context->allocated_buf, 0, context->lines_in_buf * context->scanline_width);
+    context->last_rendered_frame = lottie_animation_get_totalframe(animation);
+    context->top = 0;
+    dsc->user_data = context;
+    dsc->decoder = decoder;
+    dsc->img_data = NULL; /* We want the renderer to call decoder_read_line */
```

Comment:
If a large enough buffer is provided it can render the image here too.

Reply:
Does not really make sense anymore. If the buffer is large enough, decoder_read_line will cache the whole picture by itself (it'll not redecode the picture). So I'm not sure it's worth the trouble.

Reply:
But LVGL will still draw the image line by line which is slower.

Reply:
It draws line by line in all cases. The length of the iteration is longer when you need to call decoder_read_line, but inside the function, it's mainly doing:
```cpp
if (current line is in previous rendering) 
   color_convert
   memcpy
   return
```
Blend map is doing this:
```cpp
   (color_convert is done once at decoding time)
   memcpy
```
So yes, you've one less test, but is it really slower ?
I can rework the decoder_read_line to color convert the whole buffer upon filling it (so the latency when it's decoding for the first time will be a bit higher). That way it'll resume to a test + memcpy like the blend map.

Reply:
Well, it **currently** draw line by line in blend_map. In your new draw architecture branch, maybe it's different, I don't know. Please confirm. 
If I fill the `img_data`'s field, then I also need to allocate it and free it. I don't see where I can do that, should it be done upon closing the decoder ?

Reply:
The only loss it overhead of starting to draw let's say 20 images instead of one. (calculating areas, setting up buffers, etc). Probably not that much relative to rlottie decoding. 

The other thing we loss with read_line is that the object can't be transformed (rotated, or scaled) but I don't think it's a big issue here.

BTW, have you done a performance comparison with the original and the new decoding?

Reply:
> BTW, have you done a performance comparison with the original and the new decoding?

No, I haven't. On the simulator, it's too similar to spot (no benchmark here, just looking at the reported CPU usage). On my ESP32, the original does not fit the limited memory I have, so I can't tell.

Reply:
Ok, I've found a way to fill `img_data` when the entire picture can fit the buffer. It allows rotating the picture. So I consider this done.